napari/components/_dims/model.py

```diff
@@ -1,143 +1,297 @@
 
 
-class Dims:
-    """Dimensions object containing the dimension sliders
+class Dims():
+    """Dimensions object modeling multi-dimensional slicing, cropping, and
+    displaying in Napari
 
     Parameters
     ----------
-    viewer : Viewer
-        Parent viewer.
+    init_ndims : int, optional
+        Initial number of dimensions
 
     Attributes
     ----------
-    indices : list
-        Slicing indices controlled by the sliders.
-    max_dims : int
-        Maximum dimensions of the contained layers.
-    max_shape : tuple of int
-        Maximum shape of the contained layers.
+    events : EmitterGroup
+        Event emitter group
+    range : list
+        List of tuples (min, max, step), one for each dimension
+    point : list
+        List of floats, one for each dimension
+    interval : list
+        List of tuples (min, max), one for each dimension
+    mode : list
+        List of DimsMode, one for each dimension
+    display : list
+        List of bool indicating if dimension displayed or not, one for each
+        dimension
+    ndims : int
+        Number of dimensions
+    displayed : np.ndarray
+        Array of the displayed dimensions
     """
-    def __init__(self, viewer):
+    def __init__(self, init_ndims=0):
+        super().__init__()
 
-        self.viewer = viewer
-        self.events = EmitterGroup(source=self,
-                                   auto_connect=True,
-                                   update_slider=Event)
-        # TODO: allow arbitrary display axis setting
-        # self.y_axis = 0  # typically the y-axis
-        # self.x_axis = 1  # typically the x-axis
+        # Events:
+        self.events = EmitterGroup(source=self, auto_connect=True, axis=None,
+                                   ndims=None)
 
-        # TODO: wrap indices in custom data structure
-        self.indices = [slice(None), slice(None)]
+        self.range = []
+        self.point = []
+        self.interval = []
+        self.mode = []
+        self.display = []
 
-        self._max_dims = 0
-        self._max_shape = tuple()
+        self._add_axes(init_ndims - 1)
 
-        # update flags
-        self._child_layer_changed = False
-        self._need_redraw = False
-        self._need_slider_update = False
+    def __str__(self):
+        return "~~".join([str(self.range),
+                         str(self.point),
+                         str(self.interval),
+                         str(self.mode),
+                         str(self.display)])
 
-        self._recalc_max_dims = False
-        self._recalc_max_shape = False
+    @property
+    def ndims(self):
+        """Returns the number of dimensions
+
+        Returns
+        -------
+        ndims : int
+            Number of dimensions
+        """
+        return len(self.point)
 
-        self._qt = QtDims(self)
+    @ndims.setter
+    def ndims(self, ndims):
+        if self.ndims < ndims:
+            self._add_axes(ndims - 1)
```

Comment:
Why is this `ndims - 1` instead of `ndims`? I feel like this warrants a comment.

Reply:
I'm fixing this